### PR TITLE
tinyxxd: init at 1.3.5, use as default xxd

### PR DESCRIPTION
**Pull request by philiptaron**
## Motivation for change

- https://github.com/NixOS/nixpkgs/pull/335213
- https://github.com/NixOS/nixpkgs/pull/335269

Vim is a complex and vulnerability-prone beast. `xxd` is just a little guy. Sad to couple the two of them. So `tinyxxd` was invented by @xyproto. Let's use it.

## Description of changes

- [x] Make `tinyxxd`
- [x] Make `tinyxxd.xxd` (which exposes `tinyxxd` as `xxd` for drop-in purposes)
- [x] Document it
- [x] Yolo it [just like Arch does](https://gitlab.archlinux.org/archlinux/packaging/packages/tinyxxd/-/blob/main/PKGBUILD?ref_type=heads)

@emilazy told me to put `xxd` in the main derivation ~but I didn't do that~ so I did that.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (emilazy) [inline nixos/doc/manual/release-notes/rl-2411.section.md:371]: This is a pretty inside‐baseball change and the two should be 1:1 replacements. I don’t think we should clutter the release notes with this. (Plus, we `inherit` it from `unixtools`, so isn’t `pkgs.xxd` just as canonical?)

File: pkgs/by-name/ti/tinyxxd/package.nix
Comment (emilazy) [inline pkgs/by-name/ti/tinyxxd/package.nix:27]: I really do think that the additional output here adds no value; the only reason to separate out aliases like this is if you’re worried about polluting a shared global namespace, which isn’t a concern for Nix. I’d prefer to match what the upstream‐maintainer‐maintained Arch package does here and just `ln -s` directly into `$out`. (Who is going to think “I really want `tinyxxd` but not, like, as `xxd`”? I struggle to imagine the guy. Having `pkgs.tinyxxd` as distinct from `pkgs.xxd` but actually they’re the same – it just complicates the story, in my opinion. The core function of this piece of software is to provide an `xxd` you can use.)

If you have an intense strength of feeling about this, though, then I’ll leave it at this grumble.

File: pkgs/by-name/ti/tinyxxd/package.nix
Comment (emilazy) [inline pkgs/by-name/ti/tinyxxd/package.nix:32]: This needs to be `lib.licenses.gpl2Only` or `lib.licenses.gpl2Plus`. The Arch package implies the latter, but the repository seems to mostly imply the former. It’s also lacking a credit for the original author Juergen Weigert. I will raise these concerns upstream and we should resolve them before merging this.

File: pkgs/top-level/unixtools.nix
Comment (emilazy) [inline pkgs/top-level/unixtools.nix:229]: @rhelmot Could you check that this builds and runs on FreeBSD? (Thanks for all your work on NixBSD; I really want to try it out sometime.)

**Review by emilazy — CHANGES_REQUESTED** [review]
I’ve confirmed that this builds and works on `x86_64-darwin` (after rebasing locally on `master` – I recommend basing `staging`‐targeting PRs on `master` when possible, because it makes them easier to test). Since this package is critical to a lot of builds, before merging we should confirm it works on all the other main platforms but also with musl, LLVM, and the static package set.

tinyxxd doesn’t ship the `xxd.1` manual page that the Vim distribution does. I’ll ask upstream about that. (Hopefully that wouldn’t stop it being tiny.)

Also – I want to take a look at the diff between this code and Vim’s `xxd.c`. You talked about VIm being a vulnerability‐prone thing, but, uh, apparently that [also applies](https://github.com/vim/vim/commit/00221487731ea1868c57259c7aa0eb713cd7ade7) to [this tool](https://github.com/vim/vim/commit/67797191e039196128c69ba1538ccaf2a4711323). So I’d like to check if that stuff has been mitigated and if upstream keeps on top of Vim’s changes. (Worst case, we can always just have a separate package that talks Vim’s build system into making us an `xxd` and nothing else. But it’d be nice to share the work here and use this fork, if we can.)

File: pkgs/top-level/unixtools.nix
Comment (rhelmot) [reply pkgs/top-level/unixtools.nix:229]: It works!

**Comment by xyproto** [comment]
Hi, upstream for `tinyxxd` here. 👋 

I would gladly add a man page and also credit all past contributors. No credits have been intentionally dropped. I was mostly focused on making the fork small, functional and fast, as well as being a drop-in replacement for `xxd`.

I am happy that `tinyxxd` is being considered for Nix, since leaner build-time dependencies helps us all.

Please @ me or raise issues/PRs in the `tinyxxd` project and I'll do my best to make everything excellent.

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (philiptaron) [reply nixos/doc/manual/release-notes/rl-2411.section.md:371]: I mostly copied this from https://github.com/NixOS/nixpkgs/pull/254766, which had an even more inside baseball release note.

This note I think _does_ need to be specified, because _I_ would be surprised for about 45 seconds if `xxd` wasn't connected to Vim.

File: pkgs/by-name/ti/tinyxxd/package.nix
Comment (philiptaron) [reply pkgs/by-name/ti/tinyxxd/package.nix:27]: I like that, as currently modeled, if you ask for `tinyxxd`, you get `tinyxxd`, and if you ask for `xxd`, you get something that fulfills the role of `xxd`. The symmetry with `vim.xxd` is also (IMO) great.

I agree that basically nobody is going to install `tinyxxd`, and that everybody will use the redirector in `unixtools`. _That's_ where the logic that you're talking about -- I want an xxd -- really holds.

I'll stew on it. Still in the boat of keeping it as is though.

**Comment by philiptaron** [comment]
> Hi, upstream for tinyxxd here. 👋

Thank you for your work @xyproto! I've opened three issues in the `tinyxxd` repo related to the feedback here.

- https://github.com/xyproto/tinyxxd/issues/6
- https://github.com/xyproto/tinyxxd/issues/7
- https://github.com/xyproto/tinyxxd/issues/8

Happy to see you here. We're _completely_ aligned on making the build-time dependencies leaner.

**Comment by emilazy** [comment]
Thanks @xyproto for this package 💜 I’ll definitely bring my remaining concerns upstream and it’s always great when we can collaborate across distros.

**Comment by philiptaron** [comment]
I built:

- [x] `pkgsLLVM.tinyxxd`
- [x] `pkgsMusl.tinyxxd`
- [x] `pkgsi686Linux.tinyxxd`
- [x] `pkgsZig.tinyxxd`

Everything came up roses.

**Comment by emilazy** [comment]
We should perhaps think about what to do about Vim’s `xxd`. We probably at least want to set the priorities so that this `xxd`, which we like more, takes precedence over Vim’s.

But also maybe we should just drop the `xxd` from Vim or at least `passthru = { inherit xxd; }`? Not sure. No in‐tree packages seem to reference `pkgs.vim.xxd` so it may be largely academic, though it’s a bit sad (albeit admittedly basically irrelevant) to think that users are often going to get a less‐preferred implementation.

**Comment by philiptaron** [comment]
The other way to go is to put this xxd into the stdenv/bootstrap/thousand rebuild situations but leave Vim as the provider for pkgs.xxd.

**Comment by emilazy** [comment]
Eh, AFAICT this `xxd` is strictly better in terms of performance, testing, and code cleanliness. So I’d rather do nothing at all than that. (But I guess my most preferred option is just to set the priorities so that this wins when `pkgs.vim` and `pkgs.xxd` are both in an environment? And then we can figure out if we even want Vim to ship `xxd` later.)

**Comment by philiptaron** [comment]
I _think_ that's [`lib.meta.lowPrio`](https://noogle.dev/f/lib/meta/lowPrio) but there are likely other ways to accomplish the same end.

**Comment by emilazy** [comment]
I believe we’d want to do something like this:

```
emily@yuyuko ~/D/nixpkgs ((ac547f3d))> rg 'priority = .*\.priority.*[-+]' | cat
pkgs/desktops/lomiri/applications/lomiri-system-settings/wrapper.nix:    priority = (lomiri-system-settings-unwrapped.meta.priority or 0) - 1;
pkgs/data/fonts/powerline-symbols/default.nix:    priority = (powerline.meta.priority or 0) + 1;
pkgs/tools/networking/inetutils/default.nix:    priority = (util-linux.meta.priority or 5) + 1;
pkgs/tools/networking/openssh/copyid.nix:    priority = (openssh.meta.priority or 0) - 1;
pkgs/tools/networking/maubot/wrapper.nix:    meta.priority = (unwrapped.meta.priority or 0) - 1;
pkgs/applications/window-managers/wayfire/wrapper.nix:    priority = (wayfire.meta.priority or 0) - 1;
pkgs/applications/editors/neovim/wrapper.nix:      priority = (neovim-unwrapped.meta.priority or 0) - 1;
pkgs/applications/editors/kakoune/wrapper.nix:    meta = kakoune.meta // { priority = (kakoune.meta.priority or 0) - 1; };
pkgs/development/r-modules/wrapper.nix:      priority = (R.meta.priority or 0) - 1;
pkgs/development/r-modules/wrapper-radian.nix:    priority = (radian.meta.priority or 0) - 1;
pkgs/development/tools/build-managers/gradle/default.nix:        priority = (gradle.meta.priority or 0) + 1;
pkgs/development/tools/analysis/rizin/wrapper.nix:    priority = (unwrapped.meta.priority or 0) - 1;
pkgs/applications/networking/browsers/firefox/wrapper.nix:        priority = (browser.meta.priority or 0) - 1; # prefer wrapper over the package
pkgs/applications/networking/cluster/helm/wrapper.nix:      priority = (helm.meta.priority or 0) - 1;
pkgs/development/python-modules/dnf4/wrapper.nix:    priority = (dnf4-unwrapped.meta.priority or 0) - 1;
pkgs/applications/misc/rofi/wrapper.nix:    priority = (rofi-unwrapped.meta.priority or 0) - 1;
```

Not sure whether it should go in `vim` or `tinyxxd`.

**Comment by philiptaron** [comment]
I'm setting this PR down for the rest of the weekend. @emilazy, you're free to push to my branch and update it to whatever you see fit in order to get it into a non-draft stage might be.

**Comment by xyproto** [comment]
I added `tinyxxd.1` (man page), `AUTHORS.md` (list of all previous and current contributors to xxd + tinyxxd), clarified that the license is currently only GPL2 (but a dual X11-MIT / GPL2 license might be possible in the future), clarified symlinking/replacing xxd in the `README.md` and even added a small `CoC.md`.

File: pkgs/by-name/ti/tinyxxd/package.nix
Comment (xyproto) [reply pkgs/by-name/ti/tinyxxd/package.nix]: Added credits.

The dual license from ViM is not 100% clear if GPL-2.0 is `GPL-2.0-only` or `GPL-2.0-or-later`.

File: pkgs/by-name/ti/tinyxxd/package.nix
Comment (SuperSandro2000) [reply pkgs/by-name/ti/tinyxxd/package.nix:30]: Multiple outputs can complicate things. We should just use the $out output here unless we have a good reason that this isn't enough.

File: pkgs/by-name/ti/tinyxxd/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/ti/tinyxxd/package.nix:29]:
```suggestion
    ln -s $out/bin/tinyxxd $out/bin/xxd
```

File: pkgs/by-name/ti/tinyxxd/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/ti/tinyxxd/package.nix:22]:
```suggestion
```

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (philiptaron) [reply nixos/doc/manual/release-notes/rl-2411.section.md:413]: I decided to keep this.

File: pkgs/by-name/ti/tinyxxd/package.nix
Comment (emilazy) [inline pkgs/by-name/ti/tinyxxd/package.nix:38]: This implies `MIT AND GPL-2.0-only`, whereas it’s actually `OR`. The two options used in‐tree here are `license = lib.licenses.mit; # OR GPL-2.0-only` (my preference, as we can choose which licence to use as a downstream, and the GPL’s requirements are a superset of MIT’s), or `license = [ lib.licenses.mit /*or*/ lib.licenses.gpl2Only ];` (somewhat incoherent IMO, if you consider how it works with a case where a package is available under e.g. either the AGPL or an unfree licence, but more popular in the tree despite this).

We really need to just encode full SPDX expressions already…

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (emilazy) [reply nixos/doc/manual/release-notes/rl-2411.section.md:413]: Fair enough, we can have the note, but I think we should drop the “canonical” sentence or at least change it to just `xxd`, because `unixtools` is somewhere between silly and an implementation detail IMO. The in‐tree users seem to generally use `xxd`, and that seems correct to me because it makes overrides easier.

**Review by emilazy — COMMENTED** [review]
Ah, you beat me to it :) LGTM in principle, a couple nits.

I’d like to add `passthru` stuff to run upstream’s tests against Vim’s `xxd` but that can wait until this is in‐tree.

File: pkgs/by-name/ti/tinyxxd/package.nix
Comment (philiptaron) [reply pkgs/by-name/ti/tinyxxd/package.nix:38]: I'll do that edit.

**Comment by philiptaron** [comment]
> I believe we’d want to do something like this:
> 
> ```
> emily@yuyuko ~/D/nixpkgs ((ac547f3d))> rg 'priority = .*\.priority.*[-+]' | cat
> pkgs/desktops/lomiri/applications/lomiri-system-settings/wrapper.nix:    priority = (lomiri-system-settings-unwrapped.meta.priority or 0) - 1;
> pkgs/tools/networking/inetutils/default.nix:    priority = (util-linux.meta.priority or 5) + 1;
> ```
> 
> Not sure whether it should go in `vim` or `tinyxxd`. (Probably the latter, actually, so we don’t interfere with e.g. `vim` vs. `neovim`.)

Agree that this priority thing should go in `tinyxxd`. 

After taking a look through the tree about priority, it's a bit confusing. I had the idea that higher priority numbers are preferred. Nope! The lower the better.

1. `nix-env` and `nix profile` are the source of the priority thing. The code is here: https://github.com/NixOS/nix/blob/b89eca9aecc69d4dfc0f0afd9353c126eb7b5858/src/nix/profile.cc#L47. Note the default priority is 5.
2. The only documentation on priority ordering I was able to find was in the source code for Nix. https://github.com/NixOS/nix/blob/b89eca9aecc69d4dfc0f0afd9353c126eb7b5858/src/nix/profile.cc#L465-L480
3. `nixpkgs` thinks that "low priority" is 10. https://github.com/NixOS/nixpkgs/blob/8843893c9b842fcac17263a5700ee496e2cbee7f/lib/meta.nix#L159
4. `buildEnv` thinks that default priority is 5. https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/buildenv/default.nix#L71
5. These `or 0` think that the default priority is 0. Since lower numbers are better, `+` or `-` for these usually don't matter: it'll be high priority since very few packages set the priority.

What a mess!

**Comment by philiptaron** [comment]
All right, I think we're in shape for this transition.

- [x] Release note with "keep calm and carry on" added.
- [x] License details finessed
- [x] `meta.priority` vs. `vim.xxd` set

I think we're ready to replace `vim.xxd` and make `vim` updates no longer have to go through staging.

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (emilazy) [inline nixos/doc/manual/release-notes/rl-2411.section.md:420]: (I don’t want to commit to providing that separate output in Vim forever, but for at least a release it should be okay.)

**Review by emilazy — APPROVED** [review]
LGTM!

**Pull request merged**
